Feature/PPM-345 create auto configuration task

morantr opened the pull request:
Create an AP Autoconfiguration task.

Passed certification tests:
https://gitlab.com/prpl-foundation/prplMesh/-/pipelines/176499531

kantera800 commented on agent/src/beerocks/slave/tasks/task.h:
```diff
@@ -48,8 +48,9 @@ class Task {
+     * @param event_obj Pointer to some chunk of memory used to pass data to the event handler.
      */
-    virtual void handle_event(uint8_t event_enum_value){};
+    virtual void handle_event(uint8_t event_enum_value, void *event_obj){};
```
lets say you have 2 ap_autoconfiguration tasks in the unified agent running in parallel , 
how do you make sure they don't write to the same "chunk of memory"
why not use AgentDB?

morantr replied:
1. There is no such thing task of the same type running in parallel on the Agent.
2. Even if it was allowed to run several instances of the same task, they would not run in parallel. They would run one after another. This is how it designed.
3. The chunk of memory is a way to pass some type of object without explicitly declare it on the handle_event signature,
the only difference is that it passed as void pointer instead of explicit type.
Either way, the incoming chuck of memory is for reading, and not writing.

kantera800 replied:
> They would run one after another. 

that's exactly what i mean and now you are allowing them to use shared resource without protecting it.
>    Either way, the incoming chuck of memory is for reading, and not writing.

the user doesn't know its not for writing. you can't make everybody read instructions all the time.

If you must pass it as a const.

morantr replied:
This is not a shared resource, only one task will get it.

kantera800 replied:
Not true since you are adding it to the task class. 
theoretically both task could be triggered with a pointer to the same "chunk of memory"

morantr replied:
We agreed that it is a valid change in a private conversation, therefore mark as resolve.

kantera800 commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1725,7 +1725,9 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<sRadioInfo>
-            send_autoconfig_search_message(radio->front.iface_name);
+            m_task_pool.send_event(eTaskType::AP_AUTOCONFIGURATION,
+                                   ApAutoConfigurationTask::eEvent::START_AP_AUTOCONFIGURATION,
+                                   &radio->front.iface_name);
```
you should consider at least pass event data as a const .

morantr replied:
Ok, it is no harm to do it.

kantera800 commented on agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp:
```diff
@@ -68,27 +68,79 @@ void ApAutoConfigurationTask::work()
             break;
         }
         case eState::CONFIGIRED: {
+            ap_autoconfig_not_completed &= false;
```
```suggestion
            ap_autoconfig_not_completed = false;
```

morantr replied:
There is a reason I added the `&`, so if one of the radios did not finish the discovery flow, then the autoconfiguration status will not be marked as completed on the database.

kantera800 replied:
if you want indication that both radios finished discovery you should use  `db _statuses.ap_autoconfiguration_completed ` in line 155.
please avoid  `&=` and use something more code readable.

Also, i would add this (or similar) comment to the code as well
`if one of the radios did not finish the discovery flow, then the autoconfiguration status will not be marked as completed on the database`

vitalii-komisarenko replied:
Did you mean `ap_autoconfig_not_completed  = ap_autoconfig_not_completed  && false` ?
AFAIK only `&&`, `||` and `?` perform lazy evaluation in C++.

morantr replied:
I removed it completely, and replace it with a counter that is incremented on the CONFIGURED state which I compare to the number of radios. This way I also avoid a reverse logic variable name.

kantera800 commented on tests/environment.py:
```diff
@@ -65,7 +65,8 @@ def prplmesh_command(self, command: str, *args: str) -> bytes:
         '''
```
This commit is not related to this PR at all merge it as a separate fixup .
Also fix commit message typos.
`Commit 603b38e adds an option not to fail on log missmatch.`

morantr replied:
I need this commit to pass testflows on CI. Without it, it marks it as error on the test log.
fix commit message.

vitalii-komisarenko commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -192,6 +194,12 @@ backhaul_manager::backhaul_manager(const config_file::sConfigSlave &config,
+    if (!ap_auto_configuration_task) {
+        LOG(ERROR) << "failed to allocate ap_auto_configuration_task Task!";
+    }
+    m_task_pool.add_task(ap_auto_configuration_task);
```
Probably the task should not be added into the task pool if it failed to allocate.

morantr replied:
OOPS!
Fixing.
I moved the generating of the task to the init function which can fail.

vitalii-komisarenko commented on agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp:
```diff
@@ -0,0 +1,111 @@
+static constexpr uint8_t MAX_FAILED_AUTOCONFIG_SEARCH_ATTEMPTS = 20;
+static constexpr uint8_t AUTOCONFIG_DISCOVERY_TIMEOUT_SECONDS  = 1;
+
+#define FSM_MOVE_STATE(radio_iface, new_state)                                                     \
+    ({                                                                                             \
+        LOG(TRACE) << "AP_AUTOCONFIGURATION " << radio_iface                                       \
+                   << " FSM: " << fsm_state_to_string(m_state[radio_iface].state) << " --> "       \
+                   << fsm_state_to_string(new_state);                                              \
+        m_state[radio_iface].state = new_state;                                                    \
+    })
```
This kind of macro is used in multiple places. What about moving it to a common header file?

morantr replied:
It is similar but not the same. This one has unique implementation that fit only to this task.
Anyway, you can suggest opening a task on creating a common FSM infrastructure. I just don't want to do it as part of this PR.

vitalii-komisarenko commented on agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp:
```diff
@@ -0,0 +1,111 @@
+        m_state[radio_iface].state = new_state;                                                    \
+    })
+
+const std::string ApAutoConfigurationTask::fsm_state_to_string(eState status)
+{
+    switch (status) {
+    case eState::UNCONFIGURED:
+        return "UNCONFIGURED";
+    case eState::CONTROLLER_DISCOVERY:
+        return "CONTROLLER_DISCOVERY";
+    case eState::WAIT_FOR_CONTROLLER_DISCOVERY_COMPLETE:
+        return "WAIT_FOR_CONTROLLER_DISCOVERY_COMPLETE";
+    case eState::AP_CONFIGURATION:
+        return "AP_CONFIGURATION";
+    case eState::WAIT_FOR_AP_CONFIGURATION_COMPLETE:
+        return "WAIT_FOR_AP_CONFIGURATION_COMPLETE";
+    case eState::CONFIGIRED:
+        return "CONFIGIRED";
+    default:
+        LOG(ERROR) << "state argument doesn't have an enum";
+        break;
+    }
+    return std::string();
+}
```
It backhaul manager string representation is generated by `const char *backhaul_manager::s_arrStates[] = {FOREACH_STATE(GENERATE_STRING)};`.
Let's unify the code :)

morantr replied:
TBH I hate the whole "FOREACH_STATE(GENERATE_STRING)}" macro on the backhaul manager. We suppose to write code in CPP and not in MACRO.
Using macro should take place only if it has a good purpose.
The only reason I have chosen to use "FSM_MOVE_STATE" is because I wanted that on a state switch, it will print to the log file the line that the switch has occurred. If it was a regular function it would print the same line everywhere.

I'll might unify the code by removing this macro crap on the backhaul manager as part of the unified agent epic. 🙂

vitalii-komisarenko reviewed (commented):
Minor comments:
1) in commits 60dcc170e624, 1e3d28b72f4e "to Agent DB" is written twice
2) backhaul manager thread has several useful macros to add some syntax sugar for the state machine. They can be shared in a file called "state_machine_utils.h". A separate task can be created for it, since it is not urgent.

itayx commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1019,7 +1018,7 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
         db->backhaul.connection_type = AgentDB::sBackhaul::eConnectionType::Invalid;
 
-        controller_bridge_mac.clear();
+        db->controller_info.bridge_mac = {};
```
```suggestion
        db->controller_info.bridge_mac = {.oct = {0}};
```

morantr replied:
Replaced with ` db->controller_info.bridge_mac = network_utils::ZERO_MAC`

itayx reviewed (approved):
Looks good, approving but with a small fix

kantera800 commented:
please make sure you test autoconfiguration on real platform and pass at least 4.8.1 agent certification test.

vitalii-komisarenko commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2825,8 +2828,9 @@ bool backhaul_manager::handle_1905_autoconfiguration_response(ieee1905_1::CmduMe
     auto db = AgentDB::get();
-    if (!controller_bridge_mac.empty() && src_mac != controller_bridge_mac) {
-        LOG(INFO) << "current controller_bridge_mac=" << controller_bridge_mac
+    if (db->controller_info.bridge_mac != network_utils::ZERO_MAC &&
```
I just wonder if it is worth adding automatic conversion of sMacAddr to bool.
Comparing mac addresses to ZERO_MAC often happens often in the code.

morantr replied:
What is the problem with comparing `sMacAddr` to `ZERO_MAC`?
Comparing `sMacAddr` to `false` is weird IMO.

vitalii-komisarenko replied:
I mean something like this:
`if (db->controller_info.bridge_mac && (src_mac != db->controller_info.bridge_mac))`

vitalii-komisarenko reviewed (changes requested):
I agree with @kantera800 that using `&=` with booleans is unclear.
Personally I don't understand the difference between `x = false` and `x &= false`. I am also not sure that bitwise operation on booleans are guaranteed to work as expected in C++.

All other code looks good to me.

morantr committed:
agent: Move 'is_prplmesh_controller' to Agent DB

Move 'is_prplmesh_controller' to Agent DB and use it instead of local class members
on the backhaul manager and the son_slave.

On the backhaul manager, removed the print that mentions that the controller already marked as
prplmesh controller since it is irrelevant how it was marked before getting the ap_autoconfiguration
response.
Instead, reset this field to 'false' before sending the ap_autoconfiguration search message.
This is the correct thing to do.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Move 'controller_bridge_mac' to Agent DB

Move 'controller_bridge_mac' to Agent DB and use it instead of
local class members on the backhaul manager and the son_slave.

On the backhaul manager autoconfiguration_response handler, the controller
bridge mac was set twice, one time happens always, and the other time when
there is a match between the radio band type to the one received on the
response message. The second set of this parameter is redundant and
therefore removed.

The son_slave use to set the controller bridge mac on the master_socket with
"setPeerMac". This is redundant since 'getPeerMac' is not used anywhere on
the son_slave, therefore removed.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Remove "master_discovery_socket" from backhaul header file

Remove "master_discovery_socket" from the backhaul header file since it is not used anywhere.
Since the member name suggest that it related to discovery flow, it is removed as part of this PR.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Remove 'controller_discovered' flag

'controller_discovered' is used by the 'finalize_slaves_connect_state' to skip sending
a connect event to son_slave which we didn't get an Ap-Autoconfiguration response
on its band.
This is check is not important since the controller will always send responses on both
bands, and the Ap-Autoconfiguration response handler calls to the
'finalize_slaves_connect_state' on the socket of the slave that has the same band as
the received response.
In a future perspective, all the Ap-Autoconfiguration flow will be handled completely
on the Ap-Autoconfiguration task, so there is no reason to keep it.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add option to add and data object to task event

Preparative commit.
The ap_autoconfiguration task will need to get data on an incoming event.
Add to the task infrastructure ability to pass data on events.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add to the Agent DB statuses struct

Preparative commit.
Add to the Agent DB statuses struct which will be needed on one of the follow-up commits.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: ApAutoConfiguration Task skeleton and preparations

Add a skeleton for the ApAutoConfiguration Task, which includes tasks state machine and
functions that will be overridden from the Task base class.
Add the new task to eTaskType.
Add the task to the task pool on the backhaul manager.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Move send_autoconfig_search_message function to the new task

Move "send_autoconfig_search_message" function to the new ApAutoConfiguration Task.
Note that this commit break code compilation for easy code review purposes.
The compilation issue will be fixed on future commit.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add an event handler to ApAutoConfiguration Task

Add an event handler to ApAutoConfiguration Task.
Add START_AP_AUTOCONFIGURATION event to start the autoconfiguration,
and send it from the backhaul manager.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Move the SEND_AUTOCONFIG_SEARCH_MESSAGE state content to the task

Move the SEND_AUTOCONFIG_SEARCH_MESSAGE state content to the ApAutoConfiguration
task internal state machine.
To do it, replace the SEND_AUTOCONFIG_SEARCH_MESSAGE and
WAIT_FOR_AUTOCONFIG_RESPONSE_MESSAGE states with new WAIT_FOR_AUTOCONFIG_COMPLETE
that monitor the database and check if the autoconfiguration
(only discovery for now) finished successfully.

Removed from the backhaul manager class constants that are not being used anymore.

This commit returns the code to compile again, though the autoconfiguration flow is yet broken
and will be fixed on future commit.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Move ap autoconfiguration response handler

Move ap autoconfiguration response handler to the ApAutoConfiguration task,
from the backhaul manager.
Thus complete the discovery flow, and make it operational again.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
testflows: Remove error print on a non-error scenario

Commit 603b38e4117ee7a78159cc88517cb9a51f82c6c6 adds an option to not fail on log mismatch.
Even though we are not failing, the mismatched print to the log is being printed as an error in red.
Drill down the fail_on_mismatch flag to the function that prints the error, so it will not print
mismatch as an error if that is allowed.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
cppcheck: Update existing issues file

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
testflows: Tweak testflows sleeps on dev_reset_default test

Increase the sleep time by one second after configuring the backhaul
back to eth, since after moving the autoconfiguration flow to a task,
one second is sometimes not enough to wait for the autoconfig search
message, the test fails.

In addition, after dev_reset_default there is a delay between the
auto_config message to the moment, that the sockets to the son_slaves
are open.
Add a delay to make sure that the son_slaves are operational before
continuing to the next test.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

vitalii-komisarenko reviewed (approved):
Looks good to me, approved.

morantr committed:
agent: Ignore higher_layer_data_message on non-backhaul manager slaves

Commit 9ef9819a02730fd947b40aa8e321800fde2eb8ec has moved the higher_layer_data_message handler to
the son_slave.
The print and the skip condition was wrong since only one son_slave should handle this message, which is
the backhaul manager son_slave.
Fix the print and the skip condition.

PPM-345

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
